libs/symphonia/src/bin/symphtest.rs

```diff
@@ -1,65 +1,197 @@
+    },
+}
+
+#[derive(Debug, Deserialize)]
+struct NetworkLatencyParams {
+    mean_milli_sec: Vec<f64>,
+    standard_deviation: Vec<f64>,
+    loss_probability: Vec<f64>,
+}
+
+impl NetworkLatencyParams {
+    /// Calculate and return a sample from min and max on latency fields
+    fn sample(&self) -> (f64, f64, f64) {
+        let mut rng = rand::thread_rng();
+        let mean = rng.gen_range(self.mean_milli_sec[0], self.mean_milli_sec[1]);
+        let standard_deviation =
+            rng.gen_range(self.standard_deviation[0], self.standard_deviation[1]);
+        let loss_probability = rng.gen_range(self.loss_probability[0], self.loss_probability[1]);
+        (mean, standard_deviation, loss_probability)
+    }
+}
+
+#[derive(Debug, Deserialize)]
+struct ParticipantParams {
+    pareto_alpha: Vec<f64>,
+    num_participants: Vec<u64>,
+}
+
+impl ParticipantParams {
+    fn sample(&self) -> Vec<u64> {
+        // TODO: sample pareto alpha and skew the voting weight per participant
+        let mut rng = rand::thread_rng();
+        let num_participants = rng.gen_range(self.num_participants[0], self.num_participants[1]);
+        vec![100; num_participants as usize]
+    }
+}
+
+#[derive(Debug, Deserialize)]
+struct Params {
+    latency: NetworkLatencyParams,
+    participants: ParticipantParams,
 }
 
 fn main() {
+    env_logger::from_env(Env::default().default_filter_or("symphonia=trace,warn")).init();
     let opt: Opt = Opt::from_args();
-    println!("{:?}", opt);
     smol::block_on(async move {
-        env_logger::from_env(Env::default().default_filter_or("symphonia=trace,warn")).init();
-        let mut harness = Harness::new(MockNet {
-            latency_mean_ms: opt.latency_mean_ms,
-            latency_variance: opt.latency_variance,
-            loss_prob: opt.loss_prob,
-        });
-        for particpant_weight in opt.participant_weights.iter() {
-            harness =
-                harness.add_participant(tmelcrypt::ed25519_keygen().1, particpant_weight.clone());
+        match opt {
+            Opt::Rounds {
+                rounds,
+                latency_mean_ms,
+                latency_standard_deviation,
+                loss_prob,
+                participant_weights,
+            } => {
+                for _ in 0..rounds {
+                    let mock_net = MockNet {
+                        latency_mean_ms,
+                        latency_standard_deviation,
+                        loss_prob,
+                    };
+                    // TODO: avoid clone by using immutable vector conversion before loop
+                    run_round(participant_weights.clone(), mock_net).await
+                }
+            }
+            Opt::Parameterize {
+                rounds,
+                file_name,
+                test_count,
+            } => {
+                // Load file and deserialize into params
+                let mut path = env::current_dir().expect("Failed to get current directory");
```

Comment:
@nullchinchilla this is where the main code is.  I want to persist the values (rounds, contents of file-names as a Params object, test number and the actual times it took to reach consensus) into a csv output.  So I need one row per round.  Any efficient trivial way to do this? I guess I just use print! but not sure if I also need a task or some logic to detect it taking too long, etc.

Reply:
Stylistically, you may want to separate each subcommand into its own function. This also means that `Opt`'s different cases should be like 

```rust
enum Opt {
    Round(RoundOpts),
    ...
}

struct RoundOpt {...}
```

so that you can just write a function that takes in a `RoundOpt` to implement the logic for the round subcommand. See [geph4](https://github.com/geph-official/geph4/blob/master/geph4-client/src/main.rs) where I wrote:

```rust
#[derive(Debug, StructOpt)]
enum Opt {
    Connect(main_connect::ConnectOpt),
    Sync(main_sync::SyncOpt),
    BinderProxy(main_binderproxy::BinderProxyOpt),
}
```

I also think you should not call the iterations "rounds". "Rounds" has a very specific meaning in Symphonia, which is the internal number of rounds it takes to reach consensus. Consensus may take more than one round due to lost votes, etc. `Harness::run()` runs these "rounds" repeatedly until consensus is reached. Instead, call it "iterations" or some other term that avoids confusion.

Finally, to answer the question, the best way is probably to create some sort of `StatGatherer` struct and modify `Harness::run` to take a `&mut StatGatherer` and stuff stats into it as it goes along.

To detect whether `Harness::run` is taking too long, you can use the `smol-timeout` crate, or just roll it manually, like:

```rust
use smol::prelude::*;

...
let success_fut = async {
    harness.run().await;
    true
}; // impl Future<Output = bool>
let fail_fut = async {
    smol::Timer::after(Duration::from_secs(60)).await;
    false
}; // impl Future<Output = bool>
let succeeded = success_fut.or(fail_fut).await; // bool 
```

Async Rust lets you do these cool future-combining tricks without even spawning tasks.